src/main/groovy/lang/ObjectRange.java

```diff
@@ -368,149 +367,152 @@ public String inspect() {
      * Also see containsWithinBounds.
      */
     public boolean contains(Object value) {
-        Iterator it = new StepIterator(this, 1);
-        if (value == null) return false;
-        while (it.hasNext()) {
-            try {
-                if (DefaultTypeTransformation.compareEqual(value, it.next())) return true;
-            } catch (ClassCastException e) {
-                return false;
+        if (value == null)
+            return false;
+
+        for (Iterator it = iterator(); it.hasNext(); ) {
+            if (compareEqual(value, it.next())) {
+                return true;
             }
         }
         return false;
     }
 
+    @Override
     public void step(int step, Closure closure) {
-        if (step == 0 && compareTo(from, to) == 0) {
+        step((Number) step, closure);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void step(Number step, Closure closure) {
+        if (compareEqual(step, 0) && compareTo(from, to) == 0)
             return; // from == to and step == 0, nothing to do, so return
-        }
-        StepIterator iter = new StepIterator(this, step);
-        while (iter.hasNext()) {
-            closure.call(iter.next());
-        }
+
+        for (StepIterator it = new StepIterator(this, step); it.hasNext(); )
+            closure.call(it.next());
     }
 
     /**
      * {@inheritDoc}
      */
     public Iterator iterator() {
-        // non thread-safe iterator
-        final Iterator innerIterator = new StepIterator(this, 1);
-        Iterator safeIterator = new Iterator() {
+        return new Iterator() {
+            final Iterator innerIterator = new StepIterator(ObjectRange.this, 1);
+
             @Override
-            public synchronized boolean hasNext() {
+            public boolean hasNext() {
```

Comment:
The innerIterator is not thread-safe, meaning if two threads access next() in parallel, they may get junk. By synchronizing the wrapper, only one thread at a time gets through to the inner interator, other threads are blocked in the wrapper until the thread inside is done. That way, this method returns a thread-safe iterator.